### ISSUE TITLE
[Bug] Input validation and logical inconsistency in MeanIoU segmentation metrics

### ISSUE THREAD

**Issue by amadhan882** [keras-team/keras#22835]
## Description
`MeanIoU` and related segmentation metrics lack input validation for class IDs, causing backend crashes when labels exceed the pre-allocated confusion matrix size. Additionally, the metric returns a score of approximately 1.0 for classes absent from both ground truth and predictions, which artificially inflates results.

## Reproduction Script

```python
import os
os.environ["KERAS_BACKEND"] = "tensorflow"
import keras
import numpy as np

try:
    m_bound = keras.metrics.MeanIoU(num_classes=2)
    # ID 5 exceeds matrix size [2, 2]
    m_bound.update_state([5], [5]) 
except Exception as e:
    print(f"Caught Error: {e}")

m_nan = keras.metrics.IoU(num_classes=2, target_class_ids=[0, 1])

m_nan.update_state([0], [0]) 
print(f"Zero-instance result: {m_nan.result().numpy()}")
```

## Observed behavior
```
Caught Error: {{function_node __wrapped__ScatterNd_device_/job:localhost/replica:0/task:0/device:GPU:0}} indices[0] = [5, 5] does not index into shape [2,2]
	 [[{{node ScatterNd}}]] [Op:ScatterNd]
Zero-instance result: 0.9999997615814209

```

## Expected behavior
* The metric should validate that input labels are within the range `[0, num_classes - 1]` and raise a `ValueError` if they are not.
* Classes with zero instances should logically contribute 0.0 or be excluded from the mean calculation.

## Notes
The lack of API-level validation leads to ungraceful crashes. Furthermore, the current handling of missing classes creates misleadingly high performance scores in sparse segmentation tasks.

**Comment by rahulrathnavel** [comment]
Hi @dhantule ,
I have opened PR #22844 to address this issue. I added the bounds validation mask to safely filter out invalid class IDs and fixed the absent-class division logic. All multi-backend tests and formatting checks are passing!

So kindly let me know if you need any adjustments during your review.

**Comment by dhantule** [comment]
Hi @amadhan882, thanks for reporting this. 

I have tested this issue with `Keras 3.16.0`, the backend crash no longer reproduces, so that part appears to be fixed. 
However, the IoU metric still returns approximately `1.0` for the zero-instance case, so that behavior still appears to be present, attaching a [gist](https://colab.sandbox.google.com/gist/dhantule/314b64af614c631d8d5490ac01c4a850/-bug-input-validation-and-logical-inconsistency-in-meaniou-segmentation-metrics-22835.ipynb).

**Comment by Almond922** [comment]
Hi @dhantule, thanks for confirming the remaining issue and for the repro gist — really helpful.

I'd like to pick up the fix for the zero-instance class behavior (classes absent from both ground truth and predictions returning ~1.0 instead of 0.0). I'll take a look at the IoU/MeanIoU implementation in keras/src/metrics and put together a PR that either returns 0.0 for zero-instance classes or excludes them from the mean, per the original expected behavior in the issue description.

Let me know if there's a preferred approach between those two (return 0.0 vs. exclude from mean), or if there's already context/discussion on which is preferred.
